samtranslator/model/sam_resources.py

```diff
@@ -2300,9 +2303,49 @@ def _construct_ddb_datasources(
                 cfn_datasource.Description = cast(PassThrough, ddb_datasource.Description)
 
             cfn_datasource.DynamoDBConfig = self._parse_ddb_config(ddb_datasource)
-            datasources.append(cfn_datasource)
 
-        return datasources
+            cfn_datasource.ServiceRoleArn, permissions_resources = self._parse_datasource_role(
+                ddb_datasource, cfn_datasource.get_runtime_attr("arn"), relative_id, kwargs
+            )
+
+            resources.extend([cfn_datasource, *permissions_resources])
+
+        return resources
+
+    def _parse_datasource_role(
+        self,
+        ddb_datasource: aws_serverless_graphqlapi.DynamoDBDataSource,
+        datasource_arn: Intrinsicable[str],
+        relative_id: str,
+        kwargs: Dict[str, Any],
+    ) -> Tuple[str, List[Resource]]:
+        # If the user defined a role, then there's no need to generate role/policy for them, so we return fast.
+        if ddb_datasource.ServiceRoleArn:
+            return cast(PassThrough, ddb_datasource.ServiceRoleArn), []
+
+        # If the user doesn't have their own role, then we will create for them if TableArn is defined.
+        table_arn = sam_expect(
+            ddb_datasource.TableArn, relative_id, f"DynamoDBDataSources.{relative_id}.TableArn"
+        ).to_not_be_none("'TableArn' must be defined to create the role and policy if 'ServiceRoleArn' is not defined.")
+        permissions = ddb_datasource.Permissions or ["Read", "Write"]
+
+        role_id = f"{relative_id}Role"
+        role = IAMRole(
+            logical_id=role_id,
+            depends_on=self.depends_on,
+            attributes=self.resource_attributes,
+        )
+        role.AssumeRolePolicyDocument = IAMRolePolicies.construct_assume_role_policy_for_service_principal(
+            "appsync.amazonaws.com"
+        )
+        role.RoleName = role_id
```

Comment:
We shouldn't set names (especially global ones such as for IAM) unless necessary. This means two stacks deploying the same template will fail due to naming conflict. Instead we should pass an intrinsic that resolves to whatever CloudFormation decides.